hack/generator/pkg/config/configuration.go

```diff
@@ -0,0 +1,124 @@
+	TypeFilters []*TypeFilter
+	// TypeTransformers used to remap types
+	TypeTransformers []*TypeTransformer
+}
+
+// ShouldExportResult is returned by ShouldExport to indicate whether the supplied type should be exported
+type ShouldExportResult string
+
+const (
+	// Export indicates the specified type should be exported to disk
+	Export ShouldExportResult = "export"
+	// Skip indicates the specified type should be skipped and not exported
+	Skip ShouldExportResult = "skip"
+)
+
+// ShouldPruneResult is returned by ShouldPrune to indicate whether the supplied type should be exported
+type ShouldPruneResult string
+
+const (
+	// Include indicates the specified type should be included in the type graph
+	Include ShouldPruneResult = "include"
+	// Prune indicates the type (and all types only referenced by it) should be pruned from the type graph
+	Prune ShouldPruneResult = "prune"
+)
+
+// NewConfiguration is a convenience factory for Configuration
+func NewConfiguration(filters ...*ExportFilter) *Configuration {
+	result := Configuration{
+		ExportFilters: filters,
+	}
+
+	return &result
+}
+
+// Initialize checks for common errors and initializes structures inside the configuration
+// which need additional setup after json deserialization
+func (config *Configuration) Initialize() error {
+	if config.SchemaURL == "" {
+		return errors.New("SchemaURL missing")
+	}
+
+	for _, transformer := range config.TypeTransformers {
+		err := transformer.Init()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ShouldExport tests for whether a given struct should be exported
+// Returns a result indicating whether export should occur as well as a reason for logging
+func (config *Configuration) ShouldExport(typeName *astmodel.TypeName) (result ShouldExportResult, because string) {
+	for _, f := range config.ExportFilters {
+		if f.AppliesToType(typeName) {
+			switch f.Action {
+			case ExportFilterActionExclude:
+				return Skip, f.Because
+			case ExportFilterActionInclude:
+				return Export, f.Because
+			default:
+				panic(fmt.Errorf("unknown exportfilter directive: %s", f.Action))
+			}
+		}
+	}
+
+	// By default we export all types
+	return Export, ""
+}
+
+// ShouldPrune tests for whether a given type should be processed or pruned
+func (config *Configuration) ShouldPrune(typeName *astmodel.TypeName) (result ShouldPruneResult, because string) {
+	for _, f := range config.TypeFilters {
+		if f.AppliesToType(typeName) {
+			switch f.Action {
+			case TypeFilterPruneType:
+				return Prune, f.Because
+			case TypeFilterIncludeType:
+				return Include, f.Because
+			default:
+				panic(fmt.Errorf("unknown typefilter directive: %s", f.Action))
+			}
+		}
+	}
+
+	// By default we export all types
+	return Include, ""
+}
+
+// TransformType uses the configured type transformers to transform a type name (reference) to a different type.
+// If no transformation is performed, nil is returned
```

Comment:
Returning the original type if there is no transformation might make the consuming code simpler.

Reply:
I originally had actually done that but moved to this model instead because I felt like this was simpler - either way the caller needs to do a check (at least currently if we want logging), so the two options end up being:


```
	transformation, because := scanner.configuration.TransformType(typeName)
	if transformation != nil {
		klog.V(2).Infof("Transforming %s -> %s because %s", typeName, transformation, because)
		return transformation, nil
	}
```
vs
```
	transformation, because := scanner.configuration.TransformType(typeName)
	if transformation != typeName {
		klog.V(2).Infof("Transforming %s -> %s because %s", typeName, transformation, because)
		return transformation, nil
	}
```

I think you're thinking we could get to something like this:
```
	typeName, because := scanner.configuration.TransformType(typeName)
        // code that uses typename, no check required
```
but unfortunately the `typeName` is a more specific type than `ast.Type`, so the types don't work out (and that also makes the `transformation != typeName` check have more congnative load because you have do the "oh yea, typeName is a type too" song and dance). Given that the types don't work out to make it a one-liner, and we need to do a check for logging anyway, I think checking against nil (which is quite common/idiomatic in go) is clearer (or at least equally as clear) here.

Reply:
Actually, it might be possible to do as you say (or at least get closer to doing as you say) if `PrimitiveType` was just a `TypeName`. Part of the reason why the return of the transform function is `astmodel.Type` is because it can be a primitive type, OR a type name, and the only relationship between those two today is `astmode.Type`

If you squint and look at it, a `PrimitiveType` really is just a reference to a type defined by the standard library/language, so maybe we could do away with it entirely - will bring it up in teams chat where we can discuss it more.

Reply:
> If you squint and look at it, a `PrimitiveType` really is just a reference to a type defined by the standard library/language, so maybe we could do away with it entirely - will bring it up in teams chat where we can discuss it more.

Works for me 🙂